flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/SchedulingPipelinedRegionComputeUtil.java

```diff
@@ -0,0 +1,142 @@
+                buildRawRegions(topologicallySortedVertexes, resultPartitionRetriever);
+
+        return mergeRegionsOnCycles(
+                vertexToRegion,
+                (currentRegion, regionIndices) ->
+                        buildOutEdgesForRegion(
+                                currentRegion,
+                                regionIndices,
+                                vertexToRegion,
+                                executionVertexRetriever));
+    }
+
+    private static Map<SchedulingExecutionVertex, Set<SchedulingExecutionVertex>> buildRawRegions(
+            final Iterable<? extends SchedulingExecutionVertex> topologicallySortedVertexes,
+            final Function<IntermediateResultPartitionID, ? extends SchedulingResultPartition>
+                    resultPartitionRetriever) {
+
+        final Map<SchedulingExecutionVertex, Set<SchedulingExecutionVertex>> vertexToRegion =
+                new IdentityHashMap<>();
+
+        // iterate all the vertices which are topologically sorted
+        for (SchedulingExecutionVertex vertex : topologicallySortedVertexes) {
+            Set<SchedulingExecutionVertex> currentRegion = new HashSet<>();
+            currentRegion.add(vertex);
+            vertexToRegion.put(vertex, currentRegion);
+
+            for (ConsumedPartitionGroup consumedPartitionGroup :
+                    vertex.getConsumedPartitionGroups()) {
+                for (IntermediateResultPartitionID consumedPartitionId : consumedPartitionGroup) {
+                    SchedulingResultPartition consumedPartition =
+                            resultPartitionRetriever.apply(consumedPartitionId);
+                    // Similar to the BLOCKING ResultPartitionType, each vertex connected through
+                    // PIPELINED_APPROXIMATE is also considered as a single region. This attribute
+                    // is called  "reconnectable". Reconnectable will be removed after FLINK-19895,
+                    // see also {@link  ResultPartitionType#isReconnectable}
+                    if (!consumedPartition.getResultType().isReconnectable()) {
+                        final SchedulingExecutionVertex producerVertex =
+                                consumedPartition.getProducer();
+                        final Set<SchedulingExecutionVertex> producerRegion =
+                                vertexToRegion.get(producerVertex);
+
+                        // check if it is the same as the producer region, if so skip the merge
+                        // this check can significantly reduce compute complexity in All-to-All
+                        // PIPELINED edge case
+                        if (producerRegion != null && currentRegion != producerRegion) {
+                            currentRegion =
+                                    mergeRegions(currentRegion, producerRegion, vertexToRegion);
+                        }
+                    } else {
+                        break;
```

Comment:
Instead of doing a break here, I prefer to check one partition from the consumedPartitionGroup in ahead, and skip looping the group if the corresponding result is not reconnect-able.

Reply:
In this way, I think we can also extract a method like `getReconnectableResults` as a param of this method and have different implementations in `SchedulingPipelinedRegionComputeUtil` and `LogicalPipelinedRegionComputeUtil`.

Reply:
In `SchedulingPipelinedRegionComputeUtil`, `producerRegion` can be null, while in `LogicalPipelinedRegionComputUtil`, `producerRegion` cannot be null. Should we merge it?

Reply:
In what case would the `producerRegion` be null?

Reply:
My mistake. The `producerRegion` cannot be null since there shouldn't be any intra-region PIPELINED edge.